### PR TITLE
doc: add design doc for stats lru cache

### PR DESCRIPTION
**Pull request by Yisaer**
Signed-off-by: yisaer <disxiaofei@163.com>

### What problem does this PR solve?

Issue Number: ref https://github.com/pingcap/tidb/issues/34052

Problem Summary:

### What is changed and how it works?

add design doc for stats lru cache

### Check List

Tests 

- [ ] No code

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: docs/design/2022-08-02-stats-lru-cache.md [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+When the Stats LRU Cache memory usage exceeds the quota, the LRU Cache will select one column or index stats to evict the data structures by following rules to reduce the memory usage:
+
+- If the status is `AllLoaded`, it will discard the `CMSketch` and the status will be changed into `OnlyCMSEvcited`
+- If the status is `OnlyCMSEvcited`, it will discard the `TopN` and the status will be changed into `OnlyHistRemained`
+- If the status is `OnlyHistRemained`, it will discard the `Histogram` and the status will be changed into `AllEvicted`
+
+### Sync Stats Compatibility
+
+Previously tidb server has Sync Stats and asynchronously Loading Histograms in order to load column stats into memory during query.
+As Stats LRU Cache Policy may evict index stats in memory, we also need Sync Stats and asynchronously Loading Histograms to support loading index stats according to its loading status to keep compatible.
```
Comment (chrysan) [inline docs/design/2022-08-02-stats-lru-cache.md:69]: Do we need more details to tell when the load will be triggered after eviction?

File: docs/design/2022-08-02-stats-lru-cache.md [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+## Background
+
+Previously, tidb maintained the all the indices' stats and some columns' stats which is needed during query. 
+As the maintained stats grows, the total memory usage of the stats will increase and makes tidb server OOM.
+Thus we use lru to maintain stats cache in order to keep memory safe.
+
+### Goal
+
+- Use LRU to maintain the stats cache in memory
+- Keep the total memory usage of stats cache under the quota
```
Comment (chrysan) [inline docs/design/2022-08-02-stats-lru-cache.md:19]: The third one, provide a way to efficiently load evicted stats back into cache once needed?

File: docs/design/2022-08-02-stats-lru-cache.md [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+### Non-Goals
+
+- Considering the stats cache are in memory, we don't provide changing stats cache into LRU without restarting
+
+## Proposal
+
+### Stats Cache Interface
+
+We will provide a Stats Cache Interface which is implemented by LRU Cache and Map Cache. 
+If the tidb server didn't enable Stats LRU Cache, it will use Map Cache by default.
```
Comment (chrysan) [inline docs/design/2022-08-02-stats-lru-cache.md:30]: What about introducing the variables to turn on LRU and to set the capacity?

File: docs/design/2022-08-02-stats-lru-cache.md (hunk 1 above)
Comment (Yisaer) [reply docs/design/2022-08-02-stats-lru-cache.md:70]: updated.

File: docs/design/2022-08-02-stats-lru-cache.md (hunk 3 above)
Comment (Yisaer) [reply docs/design/2022-08-02-stats-lru-cache.md]: updated.

File: docs/design/2022-08-02-stats-lru-cache.md (hunk 2 above)
Comment (Yisaer) [reply docs/design/2022-08-02-stats-lru-cache.md:19]: updated.

File: docs/design/2022-08-02-stats-lru-cache.md (hunk 1 above)
Comment (chrysan) [reply docs/design/2022-08-02-stats-lru-cache.md:70]: LGTM for this part, while Sync Stats doesn't have design doc in this repo yet...

**Review by chrysan — APPROVED** [review]

**Review by hawkingrei — APPROVED** [review]

**Pull request merged**